### PR TITLE
Expose randomWalk parameters in config

### PR DESCRIPTION
**Pull request by thomaseizinger**
Implementation of what is suggested in #76 since I need it for a PoC. Happy to revise if it should be done differently :)

### CODE REVIEW

**Comment by thomaseizinger** [comment]
I just realized that such a change also needs to be incorporated here:

https://github.com/libp2p/js-libp2p/blob/c4cab007af5e7ec170615941230c18c893eece8a/src/config.js#L34-L40

I'll pause here then and wait for feedback to whether this feature is actually desired.

File: src/index.js [hunk 1]
```diff
@@ -116,6 +117,9 @@ class KadDHT {
      * Random walk state, default true
      */
     this.randomWalkEnabled = !options.hasOwnProperty('enabledDiscovery') ? true : Boolean(options.enabledDiscovery)
```
Comment (vasco-santos) [inline src/index.js]: let's change the line above for what you proposed in the issue

File: src/random-walk.js [hunk 2]
```diff
@@ -164,4 +164,8 @@ class RandomWalk {
   }
 }
 
+RandomWalk.prototype.defaultQueriesPerPeriod = 1
```
Comment (vasco-santos) [inline src/random-walk.js]: Please, add these to to `constants` file instead

File: src/index.js [hunk 3]
```diff
@@ -44,6 +44,7 @@ class KadDHT {
     options = options || {}
     options.validators = options.validators || {}
     options.selectors = options.selectors || {}
+    options.randomWalk = options.randomWalk || {}
```
Comment (vasco-santos) [inline src/index.js]: it would be cleaner to have the defaults here merged with what is received.

File: src/random-walk.js [hunk 4]
```diff
@@ -29,9 +29,9 @@ class RandomWalk {
    * @returns {void}
    */
   start (queries, period, timeout) {
```
Comment (vasco-santos) [inline src/random-walk.js]: Please remove the lines below and change the above for using the default parameters, as follows:

```js
start (queries = ..., )
```

I know the function was not implemented by you, but feels like a good oportunity to change this.

**Comment by thomaseizinger** [comment]
https://github.com/libp2p/js-libp2p/pull/317 just got merged :tada: 

How are defaults handled across libp2p modules?
I have the feeling it would be better if a module like `js-libp2p-kad-dht` is unopinionated about defaults and *requires* all the config options to be set.

File: src/index.js [hunk 5]
```diff
@@ -115,7 +116,10 @@ class KadDHT {
     /**
      * Random walk state, default true
      */
-    this.randomWalkEnabled = !options.hasOwnProperty('enabledDiscovery') ? true : Boolean(options.enabledDiscovery)
+    this.randomWalkEnabled = !options.randomWalk.hasOwnProperty('enabled') ? c.defaultRandomWalk.enabled : Boolean(options.randomWalk.enabled)
+    this.randomWalkQueriesPerPeriod = !options.randomWalk.hasOwnProperty('queriesPerPeriod') ? c.defaultRandomWalk.queriesPerPeriod : parseInt(options.randomWalk.queriesPerPeriod)
+    this.randomWalkInterval = !options.randomWalk.hasOwnProperty('interval') ? c.defaultRandomWalk.interval : parseInt(options.randomWalk.interval)
+    this.randomWalkTimeout = !options.randomWalk.hasOwnProperty('timeout') ? c.defaultRandomWalk.timeout : parseInt(options.randomWalk.timeout)
```
Comment (vasco-santos) [inline src/index.js]: I would recommend using [nodeutils/defaults-deep](https://github.com/nodeutils/defaults-deep#readme) in here, and in

```js
options.randomWalk = options.randomWalk || c.defaultRandomWalk
```

With the current code, if we receive a parameter as follows:

```
{
  enabled: true
}
```

The initial options object will only have that, as it exists. With my proposal, you can remove all the lines above here and default in the options above.

**Comment by vasco-santos** [comment]
@thomaseizinger I want to have the defaults here too, as this implementation may be used by other projects, and it is better to fallback to a value than to require all the parameters.

**Comment by thomaseizinger** [comment]
@vasco-santos sorry for the delay!

The code now uses `defaultsDeep`! Please have another look if you can spot anything else :)

**Comment by vasco-santos** [comment]
@thomaseizinger thanks for the updates!

Basically, we are moving our CI from jenkins to travis now. So, the master branch already is using travis, but this PR came from an older version of it. I just rebased your PR with that to use travis as well! I will have a look to the PR!

File: src/index.js [hunk 6]
```diff
@@ -46,6 +47,7 @@ class KadDHT extends EventEmitter {
     options = options || {}
     options.validators = options.validators || {}
     options.selectors = options.selectors || {}
+    options.randomWalk = defaultsDeep(options.randomWalk, c.defaultRandomWalk)
```
Comment (vasco-santos) [inline src/index.js]: Add `options.randomWalk` to the js docs above and describe the options type for it as follows:

```js
/**
 * Random walk options
 * @typedef {Object} randomWalk
 * @property {boolean} enabled - discovery enabled (default: true)
 * @property {number} queriesPerPeriod - how many queries to run per period (default: 1)
 * @property {number} interval - how often to run the the random-walk process, in milliseconds (default: 300000)
 * @property {number} timeout - how long to wait for the the random-walk query to run, in milliseconds (default: 10000)
*/
```

Also remove `options.enabledDiscovery` from the docs

**Comment by thomaseizinger** [comment]
@vasco-santos I wasn't sure how you exactly meant the docs to be fixed so I took a guess at it :D

**Comment by vasco-santos** [comment]
It was a good guess then 🎉 

Thanks for the PR @thomaseizinger

**Pull request merged**